Allow rank args (#174)
ORNL/cpp-proposals-pub: P0009/P0009.md

    changing rank or assigning different sized static extents
- add feature test macro

> - fix layout_stride constructor to be flexible with integral types of strides array

Reviewer: ```suggestion
- fix `layout_stride` constructor to be flexible with integral types of strides array
```

Author: done

> - make extents and mdspan constructors accept either rank_dynamic or rank integer args (or array of that size)

Reviewer: ```suggestion
- make `extents` and `mdspan` constructors accept either `rank_dynamic` or `rank` integer arguments (or an `array` of that size)
```

Author: done

[2]{.pnum} `extents<Extents...>` is a trivially copyable type.

> [3]{.pnum} Let $E_r$ be the $r^{th}$ element of $Extents$.

Reviewer: This is helpful; nice catch! : - )

[4]{.pnum} $E_r$ is a _dynamic extent_ if it is equal to `dynamic_extent`, otherwise $E_r$ is a _static extent_.

> For each $E_r$ equivalent to `dynamic_extent`,  

Reviewer: ```suggestion
For each $E_r$ that is a dynamic extent,
```
1. It's not clear why we need to say "equivalent" here.
2. It would be nicer to use the definition on line 1362 immediately after using it.

Author: done

the upper bound of the interval is stored in the exposition-only array `dynamic_extents_` at
`dynamic_extents_[dynamic_index(`_r_`)]`.

[5]{.pnum} If $E_r$ is a dynamic extent, let $D_r$ be the value of `dynamic_extents_[dynamic_index(`_r_`)]`.

> The $r^{th}$ interval of an `extents` is as follows: 

Reviewer: ```suggestion
The $r^{th}$ interval of an `extents` object is as follows: 
```

  
  * [5.1]{.pnum} $[0, E_r)$ if $E_r$ is a static extent,

  * [5.2]{.pnum} otherwise $[0, D_r)$. 

---

```c++
constexpr size_t dynamic_index(size_t i) noexcept; // @_exposition only_@
```

> * [7]{.pnum} *Returns*: If `i <= sizeof...(Extents)` is `true`, the number of arguments

Reviewer: ```suggestion
* [7]{.pnum} *Returns*: If `i <= sizeof...(Extents)` is `true`, then the number of arguments
```

Author: i think that was a general thing that you don't say "then" for the returns statement.

```c++
template<class... SizeTypes>
  explicit constexpr extents(SizeTypes... exts) noexcept;
```

* [5]{.pnum} *Constraints:*
    
    * [5.1]{.pnum} `(is_convertible_v<SizeTypes, size_type> && ...)` is `true`, and
    

>     * [5.2]{.pnum} `sizeof...(SizeTypes)` is `dynamic_rank()` or `rank()`.

Reviewer: ```suggestion
    * [5.2]{.pnum} `sizeof...(SizeTypes)` equals either `dynamic_rank()` or `rank()`.
```

Author: done

>     * [5.2]{.pnum} `sizeof...(SizeTypes)` equals either `dynamic_rank()` or `rank()`.

Reviewer 2: Why do not allow any value in range [dynamic_rank(), rank()). Any further extends would need to be equal to the corresponding static extend. But this is already the case for any static rank, present before last dynamic extend.

Reviewer 2: This was answered by above.

Author: This was the majority vote from the authors, only allowing either `rank_dynamic` or `rank` (just noticed I have a word swap there for `rank_dynamic`). If we did something different I think it had to be between "arguments up to last dynamic rank" and rank. It wouldn't be between "rank_dynamic" and "rank" in that case. That said I did do a test implementation of all options.

Reviewer 2: The constructor of `mdspan` (and `extends`) should be conditionally explicit if `sizeof...(Args) > rank_dynamic()` or `N > rank_dynamic()`, i.e. when there is possibility of UB, to match span (ptr, size) constructor.
Note that `N > rank_dynamic()` is same as `N == rank() && rank_dynamic() != rank()` with your precondition. Could be also spelled as `N != rank_dynamic()`

Author: yeah it should be. Doing that now.

Author: Thanks for looking at this!

> * [5]{.pnum} *Preconditions:* If `sizeof...(SizeTypes) == rank()` is `true`,

Reviewer 2: This preconditions needs to be always checked. Imagine E = extends<dyn, 1, 2, dyn, 3, 4>. Then E e(1, 1, 2, 4) requires second and third parameter to be equal to static extends. I.e. static_extends may appaer before dynamic extend.

Author: Hey, so we had an author meeting and there was pretty strong opposition to the "leave trailing arguments off" option. So we only allow either provide only the dynamic extends or provide all. If you provide only the dynamic extents there is nothing to check.

Author: Oh and we can certainly see if LEWG on the 22nd wants to vote on requiring us to do one of the other options) (hopefully while saying it can be forwarded with that change, if the vote outcome is to do something else than what we do now).

Reviewer 2: Ok, thanks. This also answers my other comment.

>              $r$ is smaller than `sizeof()...SizeTypes)`, and 

Reviewer: ```suggestion
             $r$ is smaller than `sizeof...(SizeTypes)`, and 
```

Author: done

>              $E_r$ is a _static extent_, than

Reviewer: ```suggestion
             $E_r$ is a _static extent_, then
```

Author: done

             $E_r$ equals `std::array<size_type,rank()>{exts...}[r]`.

* [6]{.pnum} *Effects:*  For all $r$ in the range `[0, rank_dynamic())`, initializes `dynamic_extents_[r]`
             with `std::array<size_type,rank()>{dynamic...}[dynamic_index_inv(r)]`.

```c++
template<class SizeType, size_t N>
constexpr extents(const array<SizeType, N>& args) noexcept;
```

* [7]{.pnum} *Constraints:*

   * [7.1]{.pnum} `is_convertible_v<SizeType, size_type>` is `true`, and

   * [7.2]{.pnum} `N` is `rank_dynamic()` or `rank()`.

> * [5]{.pnum} *Preconditions:* If `N` is `rank()`, $r$ is smaller than `N` and $E_r$ is a _static extent_, than

Reviewer: ```suggestion
* [5]{.pnum} *Preconditions:* If `N` is `rank()`, $r$ is smaller than `N`, and $E_r$ is a _static extent_, then
```

Author: done
